replifier/README.md

```diff
@@ -0,0 +1,97 @@
+# replifier
+
+The "replifier" (replicator–verifier) is responsible for:
+
+*   Creating copies of our bags in permanent storage
+*   Verifying that a bag is "correct" -- checksums match the files, no unreferenced files were included, fetch.txt URIs are consistent, and so on
+
+
+
+## How data moves through the storage service
+
+![](data_movements.png)
+
+1.  The user uploads a bag to the **user uploads bucket**.
+
+2.  The bag unpacker unpacks the user-uploaded bag into the **unpacked bags bucket**.
+
+    We want it inside our AWS account as soon as possible (so it can't be deleted under our feet).
+
+    We don't copy it to permanent storage yet (it hasn't been verified, and we don't want to taint the permanent storage with bags with bad checksums or data).
+
+3.  The bag verifier does a **standalone verification** of the unpacked bag.
+
+    This tells us whether the bag is valid, and safe to copy to permanent storage.
+
+4.  If the bag passes verification, the **bag replicator** copies the bag from the unpacked bags bucket to the permanent storage.
+
+    e.g. in the diagram above, it runs three replications:
+
+    ```
+    src -> unpacked, dst -> primary
+    src -> unpacked, dst -> replica 1
+    src -> unpacked, dst -> replica 2
+    ```
+
+5.  Once a bag has been replicated, the bag verifier does a **replicated bag verification** on each new replica.
+
+    This re-checks that the bag is valid, and confirms that it's the same as the originally unpacked bag.
+
+In theory, each of these buckets could be backed by a different storage provider.
+In practice:
+
+*   We only support user uploads from S3.
+*   We unpack bags into a bucket in S3.
+*   The replicas can be a mix of providers: currently we have one in S3, one in S3 Glacier, one in Azure.
+
+At time of writing (July 2020), our codebase has hard-coded assumptions that the user uploads and unpacked bags buckets are both in S3.
```

Comment:
Hard to believe, but we are already in August 🙂

Reply:
😱